File: Top_100_Liked_Questions/416._Partition_Equal_Subset_Sum/solution.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        sumNums = sum(nums)
        
        if sumNums % 2 == 1:
            return False

        mem = {}
        def hasSubsetSum(i, s):
            if s == 0:
                return True
            
            if s < 0:
                return False
            
            if i >= len(nums):
                return False
            
            include = mem[(i+1, s-nums[i])] if (i+1, s-nums[i]) in mem else hasSubsetSum(i+1, s-nums[i])
            
            if include:
                mem[(i+1, s-nums[i])] = True
                return True
            
            exclude = mem[(i+1, s)] if (i+1, s) in mem else hasSubsetSum(i+1, s)
            if exclude:
                mem[(i+1, s)] = True
                return True
            
            mem[(i+1, s)] = False
            return False
            
        return hasSubsetSum(0, sumNums/2)

File: Top_100_Liked_Questions/416._Partition_Equal_Subset_Sum/solution.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var canPartition = function(nums) {
    let sum = nums.reduce((x, y) => x + y);
    
    // if sum is odd number, it cannot be divided evenly
    if (sum % 2 === 1) {
        return false;
    }
    
    let hash = {};
    
    let hasSubSum = function(i, s) {
        if (s === 0) {
            return true;
        } else if (s < 0) {
            return false;
        }
        
        if (i >= nums.length) {
            return false;
        }
        
        let include = hash.hasOwnProperty([i+1,s-nums[i]]) ? hash[[i+1,s-nums[i]]] : hasSubSum(i+1, s-nums[i]);
        
        if (include) {
            hash[[i, s]] = true;
            return true;
        }
        
        let not_include = hash.hasOwnProperty([i+1,s]) ? hash[[i+1,s]] : hasSubSum(i+1, s);
        
        if (not_include) {
            hash[[i, s]] = true;
            return true;
        }

        hash[[i, s]] = false;
        return false;
    }
    
    return hasSubSum(0, sum/2);
};